### DISCUSSION TITLE
Change all self:: by static:: in internal static method calls

### DISCUSSION THREAD

**Discussion by bertoferrero** [PrestaShop/PrestaShop#41073] [Ideas]
#### Is your feature request related to a problem?
The problem is quite simple, internal static calls into the same class are made using self:: example on SpecificPrice class:
`$query_extra = self::computeExtraConditions($id_product, ((!$all_combinations)?$id_product_attribute:null), $id_customer, null);`

If I want to override computeExtraConditions method, it will not work because "self::" is telling PHP to call the method in this class, instead of the overridden method. It forces me to override each method where computeExtraConditions is called, iterating again if this new method is also static.

#### Describe the solution you'd like
Replacing this self:: by static:: the problem would be automatically solved.
`$query_extra = static::computeExtraConditions($id_product, ((!$all_combinations)?$id_product_attribute:null), $id_customer, null);`

I publish this same problem months ago but no one takes it in consideration, this problem is a great nightmare for developers.

**Comment by matks** [comment]
Hi @prestaquality you're right, `static::` is most of the times better than `self::` and most of the time when we review Pull Requests, we request the submitted code changes to use `static::` instead of `self::`

However even though new code changes are verified to make sure no more `static::` is added to PrestaShop, there's a big pile of code that exists and has been inherited from the past that does not comply with the above rule.

> I publish this same problem months ago but no one takes it in consideration

That's false, we take it into consideration 😊  but we have a thousand things to do so we can't do everything.

Think about this [tweet](https://twitter.com/adamwathan/status/1381275398646927362?s=19) from Adam Wathan, the founder of Tailwind

> One of the most stressful parts of product development (especially open source) is when people are frustrated you haven’t made the improvement they want and automatically attribute it to complacency or ignoring the community instead of just being completely buried in work to do.

It does not mean that it won't be fixed, or that we don't care. It doesn't mean that the issue is NOT important either. It just means that other issues have been considered, for whatever reason, _more_ important, so they were (or are being) tackled first. There's an immense amount of work to be done (2100 issues right now, and growing!), and tasks take time.

Unfortunately developer time is limited, so if the amount of issues opened in a certain timeframe exceeds the amount of issues that can be processed using the available developer time, tasks will simply pile up. This is common to ALL projects.

If you find that any this matter is critical for you, and it's important to have it fixed soon, then you can _invest_ into having it done. You can try doing it on your own 😉  and submit Pull Requests to replace static by self. Maybe one folder after each folder, to avoid a very big PR. The most important thing is to share that fix with everyone by submitting a Pull Request -- that's what the open source spirit is all about.

**Comment by hibatallahAouadni** [comment]
Hello @prestaquality 

Thank you for your suggestion. 
As @matks said there is no guarantee that this feature will be developed anytime soon. If this is important to you, we strongly suggest you consider [hiring a specialist](https://www.prestashop.com/en/experts) to help you create it. 

If you are willing to implement this feature into PrestaShop as a [Pull Request](https://devdocs.prestashop.com/1.7/contribute/contribute-pull-requests/), the teams will be happy to help you by ensuring that it is appropriate and coherent with the rest of the project. If that is the case, please let us know! 

PS: @matks thanks for your quick reply :superhero: 

Thank you

**Comment by bertoferrero** [comment]
Thanks, @matks and @hibatallahAouadni for your quick answer :)
I will be creating pull requests with these fixes

**Comment by PrestaEdit** [comment]
Hi there,

Is there is something to check about this, or replace `self::` by `static::` is the only things to make ?
